queue.go

```diff
@@ -0,0 +1,297 @@
+	var q queue
+	cmd := fmt.Sprintf("queue.create_tube('%s', '%s', %s)", name, cfg.getType(), cfg.toString())
+	_, err := conn.Eval(cmd, []interface{}{})
+	if err == nil {
+		q = queue{
+			name,
+			conn,
+			makeCmdMap(name),
+		}
+	}
+	return q, err
+}
+
+func getQueue(conn *Connection, name string) (Queue, error) {
+	var q queue
+	cmd := fmt.Sprintf("return queue.tube.%s ~= null", name)
+	resp, err := conn.Eval(cmd, []interface{}{})
+	if err != nil {
+		return q, err
+	}
+
+	exist := len(resp.Data) != 0 && resp.Data[0].(bool)
+
+	if exist {
+		q = queue{
+			name,
+			conn,
+			makeCmdMap(name),
+		}
+	} else {
+		err = fmt.Errorf("Tube %s does not exist", name)
+	}
+
+	return q, err
+}
+
+// Put data to queue. Returns task.
+func (q queue) Put(data interface{}) (*Task, error) {
+	return q.put(data)
+}
+
+// Put data with options (ttl/ttr/pri/delay) to queue. Returns task.
+func (q queue) PutWithConfig(data interface{}, cfg QueueOpts) (*Task, error) {
+	return q.put(data, cfg.toMap())
+}
+
+func (q queue) put(p ...interface{}) (*Task, error) {
+	var (
+		params []interface{}
+		task   *Task
+	)
+	params = append(params, p...)
+	resp, err := q.conn.Call(q.cmd["put"], params)
+	if err == nil {
+		task, err = toTask(resp.Data, &q)
+	}
+
+	return task, err
+}
+
+// The take request searches for a task in the queue.
+func (q queue) Take() (*Task, error) {
+	var params interface{}
+	if q.conn.opts.Timeout > 0 {
+		params = q.conn.opts.Timeout.Seconds()
+	}
+	return q.take(params)
+}
+
+// The take request searches for a task in the queue. Waits until a task becomes ready or the timeout expires.
+func (q queue) TakeWithTimeout(timeout time.Duration) (*Task, error) {
+	if q.conn.opts.Timeout > 0 && timeout > q.conn.opts.Timeout {
+		timeout = q.conn.opts.Timeout
+	}
+	return q.take(timeout.Seconds())
+}
+
+func (q queue) take(params interface{}) (*Task, error) {
+	var t *Task
+	resp, err := q.conn.Call(q.cmd["take"], []interface{}{params})
+	if err == nil {
+		t, err = toTask(resp.Data, &q)
+	}
+	return t, err
+}
+
+// Drop queue.
+func (q queue) Drop() error {
+	_, err := q.conn.Call(q.cmd["drop"], []interface{}{})
+	return err
+}
+
+// Look at a task without changing its state.
+func (q queue) Peek(taskId uint64) (*Task, error) {
+	resp, err := q.conn.Call(q.cmd["peek"], []interface{}{taskId})
+	if err != nil {
+		return nil, err
+	}
+
+	t, err := toTask(resp.Data, &q)
+
+	return t, err
+}
+
+func (q queue) _ack(taskId uint64) (string, error) {
+	return q.produce("ack", taskId)
+}
+
+func (q queue) _delete(taskId uint64) (string, error) {
+	return q.produce("delete", taskId)
+}
+
+func (q queue) _bury(taskId uint64) (string, error) {
+	return q.produce("bury", taskId)
+}
+
+func (q queue) _release(taskId uint64, cfg QueueOpts) (string, error) {
+	return q.produce("release", taskId, cfg.toMap())
+}
+func (q queue) produce(cmd string, p ...interface{}) (string, error) {
+	var params []interface{}
+	params = append(params, p...)
```

Comment:
I understand this construction, and why it could be used. It is just not the case.